Typing behavior change in strawberry 0.156.2

The `@strawberry.field` annotation behavior has changed at typing level between strawberry 0.156.1 and 0.156.2. It should now be a callable

## Describe the Bug

Consider this simple code:

```python
import strawberry


@strawberry.type
class Author:
    name: str


@strawberry.type
class Query:
    @strawberry.field
    async def get_authors(self) -> list[Author]:
        return [
            Author(name="Michael Crichton")
        ]
```

Running mypy on this is ok for strawberry 0.156.1.

On strawberry 0.156.2 this gives an error:
```txt
error: Argument 1 to "field" has incompatible type "Callable[[Query], Coroutine[Any, Any, List[Author]]]"; expected "Union[StrawberryResolver[<nothing>], Callable[..., <nothing>], Callable[..., Awaitable[<nothing>]], staticmethod[<nothing>], classmethod[<nothing>]]"  [arg-type]
```

This is probably due to the declaration of the callable on awaitable (https://github.com/strawberry-graphql/strawberry/pull/2528/files), since switching to field as this is ok:

```python
import strawberry


@strawberry.type
class Author:
    name: str


@strawberry.type
class Query:
    @strawberry.field()
    async def get_authors(self) -> list[Author]:
        return [
            Author(name="Michael Crichton")
        ]
```

--- a/RELEASE.md
+++ b/RELEASE.md
@@ -1,0 +1,20 @@
+Release type: patch
+
+This release fixes a regression introduce in version 0.156.2 that
+would make Mypy throw an error in the following code:
+
+```python
+import strawberry
+
+
+@strawberry.type
+class Author:
+    name: str
+
+
+@strawberry.type
+class Query:
+    @strawberry.field
+    async def get_authors(self) -> list[Author]:
+        return [Author(name="Michael Crichton")]
+```

--- a/strawberry/field.py
+++ b/strawberry/field.py
@@ -7,6 +7,7 @@ from typing import (
     Any,
     Awaitable,
     Callable,
+    Coroutine,
     Dict,
     List,
     Mapping,
@@ -39,7 +40,9 @@ T = TypeVar("T")
 _RESOLVER_TYPE = Union[
     StrawberryResolver[T],
     Callable[..., T],
-    Callable[..., Awaitable[T]],
+    # we initially used Awaitable, but that was triggering the following mypy bug:
+    # https://github.com/python/mypy/issues/14669
+    Callable[..., Coroutine[T, Any, Any]],
     "staticmethod[T]",
     "classmethod[T]",
 ]

--- a/tests/mypy/test_fields.yml
+++ b/tests/mypy/test_fields.yml
@@ -110,3 +110,22 @@
     reveal_type(i.a)
   out: |
     main:14: note: Revealed type is "builtins.str"
+
+- case: test_does_not_put_fields_with_async_resolver_in_init
+  main: |
+    import strawberry
+
+    async def resolver() -> str:
+        return "hi"
+
+    @strawberry.type
+    class Example:
+        a: str = strawberry.field(description="Example")
+        b: str = strawberry.field(resolver=resolver)
+        c: str
+
+    i = Example(a="a", c="c")
+
+    reveal_type(i.a)
+  out: |
+    main:14: note: Revealed type is "builtins.str"

